File: 05-Algorithms/code/python/seb_alg/rng.py
"""Mirrors `typescript/src/rng.ts` exactly.

The TS side is the literal, well-known mulberry32 reference implementation
using JS's signed-int32 bitwise semantics (`|0`, `>>>0`, `Math.imul`). This
port instead keeps every intermediate value as an ordinary Python
non-negative integer in [0, 2**32 - 1] and masks with `& 0xFFFFFFFF` after
every addition and multiplication. This produces bit-for-bit identical
results to the JS version because:

- XOR, OR, and right-shift produce the same bit pattern whether that
  pattern is interpreted as signed two's-complement or as unsigned -- only
  the *numeric label* attached to a bit pattern differs, not the pattern
  itself.
- Addition and multiplication modulo 2**32 are ring-homomorphic under the
  signed <-> unsigned relabeling (a signed value and its unsigned twin
  differ by a multiple of 2**32, which vanishes mod 2**32).

So performing every step as unsigned arithmetic mod 2**32 throughout
reproduces the exact same sequence of 32-bit words as JS's mix of signed
`Math.imul`/`|0` and unsigned `>>>0`/`>>>n`.
"""

from __future__ import annotations

from typing import Callable

_MASK32 = 0xFFFFFFFF


def mulberry32(seed: int) -> Callable[[], float]:
    """Returns a zero-argument function yielding successive floats in [0, 1)."""
    state = {"a": seed & _MASK32}

    def next_value() -> float:
        a = (state["a"] + 0x6D2B79F5) & _MASK32
        state["a"] = a
        t = a
        t = ((t ^ (t >> 15)) * (t | 1)) & _MASK32
        mixed = ((t ^ (t >> 7)) * (t | 61)) & _MASK32
        t = ((t + mixed) & _MASK32) ^ t
        t &= _MASK32
        return ((t ^ (t >> 14)) & _MASK32) / 4294967296.0

    return next_value

File: 05-Algorithms/code/typescript/src/rng.ts
/**
 * mulberry32: a small, fast, deterministic pseudo-random generator (public
 * domain, widely used for reproducible simulations/tests). This is the
 * canonical reference implementation, verbatim, so its bit-for-bit behavior
 * is well established. The Python port (`seb_alg/rng.py`) reproduces the
 * exact same output by performing every step as explicit unsigned 32-bit
 * modular arithmetic: JS's `|0`/`>>>0`/`Math.imul` here operate on the
 * two's-complement bit pattern of each intermediate value, and XOR, OR, and
 * (right) shift produce the *same bit pattern* regardless of whether that
 * pattern is interpreted as signed or unsigned -- only plain addition and
 * multiplication care about the interpretation, and both are congruent mod
 * 2^32 either way. So "signed int32 ops in JS" and "unsigned uint32 ops mod
 * 2^32 in Python" compute identical bit sequences from the same seed.
 * Returns a function that yields successive floats in `[0, 1)`.
 */
export function mulberry32(seed: number): () => number {
  let a = seed | 0;
  return function next(): number {
    a = (a + 0x6d2b79f5) | 0;
    let t = Math.imul(a ^ (a >>> 15), 1 | a);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}
